cluster-provision/images/vm-image-builder/fedora-sriov-lane/cloud-config

```diff
@@ -1,27 +0,0 @@
-#cloud-config
-password: fedora
-chpasswd: { expire: False }
-ssh_pwauth: yes
-write_files:
-    - path: /etc/modules-load.d/mlx5.conf
-      content: |
-        mlx5_core
-        mlx5_ib
-    - path: /etc/modules-load.d/mlx4.conf
-      content: |
-        mlx4_core
-        mlx4_ib
-    - path: /etc/modules-load.d/i40e.conf
-      content: |
-        i40e
-    - path: /etc/modules-load.d/igb.conf
-      content: |
-        igb
-runcmd:
-  - sudo dnf install -y kernel-modules-$(uname -r) qemu-guest-agent stress
-  - sudo dnf clean all
-  - sudo hostnamectl set-hostname ""
-  - sudo hostnamectl set-hostname "" --transient
-  - sudo systemctl start qemu-guest-agent
-  - sudo sed -i /users-groups/d /etc/cloud/cloud.cfg
```

Comment:
We chosen this method in order to eliminate the race condition between cloud-init setting the password and the fedora login from the tests,
now it means that the tests will set the password
see 
https://github.com/kubevirt/kubevirt/pull/6108/files#diff-da14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR682

I am less fond of this, please consider
@EdDev FYI

Reply:
I am not sure if that solves anything. Doesn't that mean that you potentially can already log in before cloud-init is done and as such you work on an OS which is not done with the setup?

Note that the consolidated image only starts the guest agent now as the last step in cloud-init (like the other image already did) which allows a race-free login. You basically wait for the guest agent first and then start your test..

Reply:
> I am not sure if that solves anything. Doesn't that mean that you potentially can already log in before cloud-init is done and as such you work on an OS which is not done with the setup?

Both methods did solve the races we encountered (David and myself), where both don't solve waiting for all cloud-init modules to be finished.
So far so good, there are no flakes anymore about the login.
If we want to wait for all modules to be finished, we can wait for cloud-init done event and for guest-agent connected event  (which we are already waiting for right?)
The races were due to the credentials recreation.
I don't think creating this binding is the right thing to do, but to create independent methods to wait for the needed modules where they are actually needed.

> 
> Note that the consolidated image only starts the guest agent now as the last step in cloud-init (like the other image already did) which allows a race-free login. You basically wait for the guest agent first and then start your test..

The guest-agent connected signal which the tests are waiting to doesn't mean the whole cloud-init is done, it has several modules.

Reply:
> The guest-agent connected signal which the tests are waiting to doesn't mean the whole cloud-init is done, it has several modules.

That can be., but I would argue that waiting for the guest-agent this way is more accurate.

Reply:
This consolidates it to doing it one way, and it can be moved later on  to another way.

Reply:
> The guest-agent connected signal which the tests are waiting to doesn't mean the whole cloud-init is done, it has several modules.

According to the  docs it is at least very very late in the process: https://cloudinit.readthedocs.io/en/latest/topics/boot.html#final: "This stage runs as late in boot as possible".

Reply:
Adding `- sudo sed -i /users-groups/d /etc/cloud/cloud.cfg`
means that we are not forced to add cloud-init
nor need to set again passwords,
nor create binding between modules,
and that we can wait for guest-agent anyway because it already runs (if we don't disable it),
cloud-init has a file that signals the finishing of it, i think we should do the other way around when consolidating,
it will just means that we need to add wait for cloud-init done where actually needed / used.

Reply:
> Adding `- sudo sed -i /users-groups/d /etc/cloud/cloud.cfg`
> means that we are not forced to add cloud-init
> nor need to set again passwords,
> nor create binding between modules,
> and that we can wait for guest-agent anyway because it already runs (if we don't disable it),
> cloud-init has a file that signals the finishing of it, i think we should do the other way around when consolidating,
> it will just means that we need to add wait for cloud-init done where actually needed / used.

You may not use user-data in all of your tests, but as far as I see you use other parts of cloud-init a lot (like network setup). You know for sure that this is done this way.

Reply:
right, but only where we do need it, and without creating any binding between modules,
all what we need to wait for is explicit.

Reply:
> right, but only where we do need it, and without creating any binding between modules,
> all what we need to wait for is explicit.

I don't understand this. It just makes everything more complicated. The mental model this way is very simple: just wait for the guest agent and most likely things are fine. I don't have to look at different contexts so much, reducing the maintenance burden. I am not sure what exactly you mean with "without creating any binding between modules". If guest-agent is ready. Things are most likely fine.

Reply:
Imho creating binding between guest agent and cloud-init, and enforcing the user to always run cloud-init in order to set password is mental burden,
keeping tests as simple and explicit / controlled is a plus.

Things are most likely fine anyhow with the approach of the sed line.
Simply adding boot finish event of cloud-init will detach the connection between the cloud-init and guest-agent
and then tests will have the option to call

1. wait for guest-agent done event (we already have in code of course)
2. wait for cloud-init done event (only where needed)

without enforcing tests to run both cloud-init and guest-agent in order to set password where only part of those 3 are needed.

Reply:
btw `sudo sed -i /users-groups/d /etc/cloud/cloud.cfg` keeping this line, means that the VM can be logged out of the box, without setting cloud-init which is a plus i think.
I don't see any reason to remove this line, whichever the solution is chosen right?
It will work both ways (unless i am missing something).

moreover it means that disabling the guest-agent means one must use cloud-init or run it manually when running the VM standalone,
I think that it is not a good practice, it would be nicer to have a standalone machine without proprietary parts, but more straightforward.

Reply:
Readded it.

Reply:
But note that I now have to remove your guest-agent check, since the guest agent will not be started anymore by default.

Reply:
Hi, why do we want to disable qemu-guest-agent by default?

Reply:
Because all other tests, except a few networking tests use the guest-agent as the readiness check.

Reply:
> But note that I now have to remove your guest-agent check, since the guest agent will not be started anymore by default.

Its orthogonal for readding sed line isnt it ?

I don't think its that the right thing to do :/ (disabling it in first place) :/

If you do so, need to check please each place where it is removed if its allowed to be removed, or that we need to start it (which is a mental burden i think?)

PS. all those fixes from David and myself started only because the login problem.

Reply:
Hey, back then I enabled qemu-guest agent by default for SRIOV test suite because there were flakes
when it was disabled and we needed to enabled it for the tests, 
Also to untangle provisioning the VM from tests suite.
I am pretty lost here on this thread.. why do we want to disable qmeu-guest agent by default again?
What does it mean for SRIOV tests? we would need to enabled it from the tests again?

Reply:
> Its orthogonal for readding sed line isnt it ?
> 
> I don't think its that the right thing to do :/ (disabling it in first place) :/
> 
> If you do so, need to check please each place where it is removed if its allowed to be removed, or that we need to start it (which is a mental burden i think?)
> 
> PS. all those fixes from David and myself started only because the login problem.

Yes. I am just consolidating it. I understand why they are there. This is just the simplest way forward until something better. As you can see right now we have complexity in multiple places:

 * having multiple images (maintenance and build download is an issue at scale and during local development)
 * having one image with multiple flows (adds complexity at provisioning and individual tests because one has to maintain the flows and has to know which test chooses wich)

Therefore I chose it. I can leave the line in but I would want to insist on the guest agent.

Reply:
> What does it mean for SRIOV tests? we would need to enabled it from the tests again?

@ormergi  I took care of that in https://github.com/kubevirt/kubevirt/pull/6108/files#diff-da14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR682. The guest agent will be started as always, but just very late when cloud-init is almost done (the same like for the other tests).

Reply:
@oshoval @ormergi  btw. don't get me wrong. I would not mind if you have your own images with your own flows, but we create big images of a few hundred megabytes which are pulled back and forth without much benefit at this stage. If needed the images can also be forked later on again.

Reply:
Please let me know if you can live with the guest-agent and cloud-init solution. If not, then I will keep the separate image for now and close the PR. That is fine too.